In [24]:
#@title Copyright 2022 Google LLC. { display-mode: "form" }
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

<table class="ee-notebook-buttons" align="left"><td>
<a target="_blank"  href="http://colab.research.google.com/github/google/earthengine-api/blob/master/python/examples/ipynb/Earth_Engine_asset_from_cloud_geotiff.ipynb">
    <img src="https://www.tensorflow.org/images/colab_logo_32px.png" /> Run in Google Colab</a>
</td><td>
<a target="_blank"  href="https://github.com/google/earthengine-api/blob/master/python/examples/ipynb/Earth_Engine_asset_from_cloud_geotiff.ipynb"><img width=32px src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" /> View source on GitHub</a></td></table>

# Cloud GeoTiff Backed Earth Engine Assets

***Note:*** *The REST API contains new and advanced features that may not be suitable for all users.  If you are new to Earth Engine, please get started with the [JavaScript guide](https://developers.google.com/earth-engine/guides/getstarted).*

Earth Engine can load images from Cloud Optimized GeoTiffs (COGs) in Google Cloud Storage ([learn more](https://developers.google.com/earth-engine/guides/image_overview#images-from-cloud-geotiffs)).  This notebook demonstrates how to create Earth Engine assets backed by COGs.  An advantage of COG-backed assets is that the spatial and metadata fields of the image will be indexed at asset creation time, making the image more performant in collections.  (In contrast, an image created through `ee.Image.loadGeoTIFF` and put into a collection will require a read of the GeoTiff for filtering operations on the collection.)  A disadvantage of COG-backed assets is that they may be several times slower than standard assets when used in computations.

To create a COG-backed asset, make a `POST` request to the Earth Engine [`CreateAsset` endpoint](https://developers.google.com/earth-engine/reference/rest/v1alpha/projects.assets/create).  As shown in the following, this request must be authorized to create an asset in your user folder.

In [25]:
import pandas as pd
from google.cloud import storage
import os
import json

## Start an authorized session

To be able to make an Earth Engine asset in your user folder, you need to be able to authenticate as yourself when you make the request.  You can use credentials from the Earth Engine authenticator to start an [`AuthorizedSession`](https://google-auth.readthedocs.io/en/master/reference/google.auth.transport.requests.html#google.auth.transport.requests.AuthorizedSession).  You can then use the `AuthorizedSession` to send requests to Earth Engine.

In [26]:
import ee
from google.auth.transport.requests import AuthorizedSession

ee.Authenticate()  #  or !earthengine authenticate --auth_mode=gcloud
session = AuthorizedSession(ee.data.get_persistent_credentials())

Enter verification code: 4/1AVHEtk4n0mDBa5dmWkobLx18UrOLmn2a30A2jJQSnB5g4iqayqLNf9MUlPg

Successfully saved authorization token.


## Get image information from gcp
Get list of images and their respective uris from the gcp data bucket


In [35]:
def list_blobs_with_prefix(bucket_name, prefix, delimiter=None):
    

    storage_client = storage.Client(project='swhm-prod')

    # Note: Client.list_blobs requires at least package version 1.17.0.
    blobs = storage_client.list_blobs(bucket_name, prefix=prefix, delimiter=delimiter)

    # Note: The call returns a response only when the iterator is consumed.
    blob_list = []
    for blob in blobs:
        blob_list.append(blob.name)

    if delimiter:
        print("Prefixes:")
        for prefix in blobs.prefixes:
            blob_list.append([prefix])
    
    return blob_list


In [36]:
BUCKET_NAME = 'live_data_layers'
blobsout = list_blobs_with_prefix(BUCKET_NAME,'raster')

In [37]:
df = pd.DataFrame(blobsout, columns=['blob_name'])
df['uri'] = 'gs://'+BUCKET_NAME+'/'+df['blob_name']
(df)

blob_name  \
0                   rasters/Age_of_Imperviousness.tif   
1                     rasters/Flow_Duration_Index.tif   
2                    rasters/HSPF_Land_Cover_Type.tif   
3               rasters/Hydrologic_Response_Units.tif   
4                          rasters/Imperviousness.tif   
5                              rasters/Land_Cover.tif   
6                                rasters/Land_Use.tif   
7                               rasters/Landcover.tif   
8                      rasters/Population_Density.tif   
9                        rasters/Precipitation_mm.tif   
10                                 rasters/Runoff.tif   
11                       rasters/Slope_Categories.tif   
12             rasters/Total_Copper_Concentration.tif   
13  rasters/Total_Kjeldahl_Nitrogen_Concentration.tif   
14         rasters/Total_Phosphorus_Concentration.tif   
15   rasters/Total_Suspended_Solids_Concentration.tif   
16                                rasters/Traffic.tif   

                                                  uri  
0   gs://live_data_layers/rasters/Age_of_Imperviou...  
1   gs://live_data_layers/rasters/Flow_Duration_In...  
2   gs://live_data_layers/rasters/HSPF_Land_Cover_...  
3   gs://live_data_layers/rasters/Hydrologic_Respo...  
4    gs://live_data_layers/rasters/Imperviousness.tif  
5        gs://live_data_layers/rasters/Land_Cover.tif  
6          gs://live_data_layers/rasters/Land_Use.tif  
7         gs://live_data_layers/rasters/Landcover.tif  
8   gs://live_data_layers/rasters/Population_Densi...  
9   gs://live_data_layers/rasters/Precipitation_mm...  
10           gs://live_data_layers/rasters/Runoff.tif  
11  gs://live_data_layers/rasters/Slope_Categories...  
12  gs://live_data_layers/rasters/Total_Copper_Con...  
13  gs://live_data_layers/rasters/Total_Kjeldahl_N...  
14  gs://live_data_layers/rasters/Total_Phosphorus...  
15  gs://live_data_layers/rasters/Total_Suspended_...  
16          gs://live_data_layers/rasters/Traffic.tif

In [38]:

file_names = [os.path.splitext(os.path.basename(file_path))[0] for file_path in df['blob_name']]
df["asset_name"] = file_names

pretty_names  = [os.path.splitext(os.path.basename(file_path))[0].replace('_', ' ') for file_path in df['blob_name']]

df["pretty_name"] = pretty_names
df

blob_name  \
0                   rasters/Age_of_Imperviousness.tif   
1                     rasters/Flow_Duration_Index.tif   
2                    rasters/HSPF_Land_Cover_Type.tif   
3               rasters/Hydrologic_Response_Units.tif   
4                          rasters/Imperviousness.tif   
5                              rasters/Land_Cover.tif   
6                                rasters/Land_Use.tif   
7                               rasters/Landcover.tif   
8                      rasters/Population_Density.tif   
9                        rasters/Precipitation_mm.tif   
10                                 rasters/Runoff.tif   
11                       rasters/Slope_Categories.tif   
12             rasters/Total_Copper_Concentration.tif   
13  rasters/Total_Kjeldahl_Nitrogen_Concentration.tif   
14         rasters/Total_Phosphorus_Concentration.tif   
15   rasters/Total_Suspended_Solids_Concentration.tif   
16                                rasters/Traffic.tif   

                                                  uri  \
0   gs://live_data_layers/rasters/Age_of_Imperviou...   
1   gs://live_data_layers/rasters/Flow_Duration_In...   
2   gs://live_data_layers/rasters/HSPF_Land_Cover_...   
3   gs://live_data_layers/rasters/Hydrologic_Respo...   
4    gs://live_data_layers/rasters/Imperviousness.tif   
5        gs://live_data_layers/rasters/Land_Cover.tif   
6          gs://live_data_layers/rasters/Land_Use.tif   
7         gs://live_data_layers/rasters/Landcover.tif   
8   gs://live_data_layers/rasters/Population_Densi...   
9   gs://live_data_layers/rasters/Precipitation_mm...   
10           gs://live_data_layers/rasters/Runoff.tif   
11  gs://live_data_layers/rasters/Slope_Categories...   
12  gs://live_data_layers/rasters/Total_Copper_Con...   
13  gs://live_data_layers/rasters/Total_Kjeldahl_N...   
14  gs://live_data_layers/rasters/Total_Phosphorus...   
15  gs://live_data_layers/rasters/Total_Suspended_...   
16          gs://live_data_layers/rasters/Traffic.tif   

                               asset_name  \
0                   Age_of_Imperviousness   
1                     Flow_Duration_Index   
2                    HSPF_Land_Cover_Type   
3               Hydrologic_Response_Units   
4                          Imperviousness   
5                              Land_Cover   
6                                Land_Use   
7                               Landcover   
8                      Population_Density   
9                        Precipitation_mm   
10                                 Runoff   
11                       Slope_Categories   
12             Total_Copper_Concentration   
13  Total_Kjeldahl_Nitrogen_Concentration   
14         Total_Phosphorus_Concentration   
15   Total_Suspended_Solids_Concentration   
16                                Traffic   

                              pretty_name  
0                   Age of Imperviousness  
1                     Flow Duration Index  
2                    HSPF Land Cover Type  
3               Hydrologic Response Units  
4                          Imperviousness  
5                              Land Cover  
6                                Land Use  
7                               Landcover  
8                      Population Density  
9                        Precipitation mm  
10                                 Runoff  
11                       Slope Categories  
12             Total Copper Concentration  
13  Total Kjeldahl Nitrogen Concentration  
14         Total Phosphorus Concentration  
15   Total Suspended Solids Concentration  
16                                Traffic

## Build the manifest

In [39]:
def get_layer_dict(asset): 
    data = json.load(open("data/rasters.json"))
    layer_dict = data[asset]
    print(asset)
    #delete dictionaries 
    try: 
        del layer_dict["layer"]
        del layer_dict["values"]
        del layer_dict["labels"]

    except KeyError:
        pass
    

    layer_dict['pretty_name'] = asset
    return layer_dict

In [40]:
df

blob_name  \
0                   rasters/Age_of_Imperviousness.tif   
1                     rasters/Flow_Duration_Index.tif   
2                    rasters/HSPF_Land_Cover_Type.tif   
3               rasters/Hydrologic_Response_Units.tif   
4                          rasters/Imperviousness.tif   
5                              rasters/Land_Cover.tif   
6                                rasters/Land_Use.tif   
7                               rasters/Landcover.tif   
8                      rasters/Population_Density.tif   
9                        rasters/Precipitation_mm.tif   
10                                 rasters/Runoff.tif   
11                       rasters/Slope_Categories.tif   
12             rasters/Total_Copper_Concentration.tif   
13  rasters/Total_Kjeldahl_Nitrogen_Concentration.tif   
14         rasters/Total_Phosphorus_Concentration.tif   
15   rasters/Total_Suspended_Solids_Concentration.tif   
16                                rasters/Traffic.tif   

                                                  uri  \
0   gs://live_data_layers/rasters/Age_of_Imperviou...   
1   gs://live_data_layers/rasters/Flow_Duration_In...   
2   gs://live_data_layers/rasters/HSPF_Land_Cover_...   
3   gs://live_data_layers/rasters/Hydrologic_Respo...   
4    gs://live_data_layers/rasters/Imperviousness.tif   
5        gs://live_data_layers/rasters/Land_Cover.tif   
6          gs://live_data_layers/rasters/Land_Use.tif   
7         gs://live_data_layers/rasters/Landcover.tif   
8   gs://live_data_layers/rasters/Population_Densi...   
9   gs://live_data_layers/rasters/Precipitation_mm...   
10           gs://live_data_layers/rasters/Runoff.tif   
11  gs://live_data_layers/rasters/Slope_Categories...   
12  gs://live_data_layers/rasters/Total_Copper_Con...   
13  gs://live_data_layers/rasters/Total_Kjeldahl_N...   
14  gs://live_data_layers/rasters/Total_Phosphorus...   
15  gs://live_data_layers/rasters/Total_Suspended_...   
16          gs://live_data_layers/rasters/Traffic.tif   

                               asset_name  \
0                   Age_of_Imperviousness   
1                     Flow_Duration_Index   
2                    HSPF_Land_Cover_Type   
3               Hydrologic_Response_Units   
4                          Imperviousness   
5                              Land_Cover   
6                                Land_Use   
7                               Landcover   
8                      Population_Density   
9                        Precipitation_mm   
10                                 Runoff   
11                       Slope_Categories   
12             Total_Copper_Concentration   
13  Total_Kjeldahl_Nitrogen_Concentration   
14         Total_Phosphorus_Concentration   
15   Total_Suspended_Solids_Concentration   
16                                Traffic   

                              pretty_name  
0                   Age of Imperviousness  
1                     Flow Duration Index  
2                    HSPF Land Cover Type  
3               Hydrologic Response Units  
4                          Imperviousness  
5                              Land Cover  
6                                Land Use  
7                               Landcover  
8                      Population Density  
9                        Precipitation mm  
10                                 Runoff  
11                       Slope Categories  
12             Total Copper Concentration  
13  Total Kjeldahl Nitrogen Concentration  
14         Total Phosphorus Concentration  
15   Total Suspended Solids Concentration  
16                                Traffic

Create json_data

In [41]:
df['json_data'] = 'na'
for index, row in df.iterrows():
    key_value = row['pretty_name']
    try: 
        json_value = json.dumps(get_layer_dict(key_value))
    except KeyError:
        json_value = "NA"
    df.at[index, 'json_data'] = json_value


Age of Imperviousness
Flow Duration Index
HSPF Land Cover Type
Hydrologic Response Units
Imperviousness
Land Cover
Land Use
Population Density
Runoff
Slope Categories
Total Copper Concentration
Total Kjeldahl Nitrogen Concentration
Total Phosphorus Concentration
Total Suspended Solids Concentration
Traffic


In [42]:
df

blob_name  \
0                   rasters/Age_of_Imperviousness.tif   
1                     rasters/Flow_Duration_Index.tif   
2                    rasters/HSPF_Land_Cover_Type.tif   
3               rasters/Hydrologic_Response_Units.tif   
4                          rasters/Imperviousness.tif   
5                              rasters/Land_Cover.tif   
6                                rasters/Land_Use.tif   
7                               rasters/Landcover.tif   
8                      rasters/Population_Density.tif   
9                        rasters/Precipitation_mm.tif   
10                                 rasters/Runoff.tif   
11                       rasters/Slope_Categories.tif   
12             rasters/Total_Copper_Concentration.tif   
13  rasters/Total_Kjeldahl_Nitrogen_Concentration.tif   
14         rasters/Total_Phosphorus_Concentration.tif   
15   rasters/Total_Suspended_Solids_Concentration.tif   
16                                rasters/Traffic.tif   

                                                  uri  \
0   gs://live_data_layers/rasters/Age_of_Imperviou...   
1   gs://live_data_layers/rasters/Flow_Duration_In...   
2   gs://live_data_layers/rasters/HSPF_Land_Cover_...   
3   gs://live_data_layers/rasters/Hydrologic_Respo...   
4    gs://live_data_layers/rasters/Imperviousness.tif   
5        gs://live_data_layers/rasters/Land_Cover.tif   
6          gs://live_data_layers/rasters/Land_Use.tif   
7         gs://live_data_layers/rasters/Landcover.tif   
8   gs://live_data_layers/rasters/Population_Densi...   
9   gs://live_data_layers/rasters/Precipitation_mm...   
10           gs://live_data_layers/rasters/Runoff.tif   
11  gs://live_data_layers/rasters/Slope_Categories...   
12  gs://live_data_layers/rasters/Total_Copper_Con...   
13  gs://live_data_layers/rasters/Total_Kjeldahl_N...   
14  gs://live_data_layers/rasters/Total_Phosphorus...   
15  gs://live_data_layers/rasters/Total_Suspended_...   
16          gs://live_data_layers/rasters/Traffic.tif   

                               asset_name  \
0                   Age_of_Imperviousness   
1                     Flow_Duration_Index   
2                    HSPF_Land_Cover_Type   
3               Hydrologic_Response_Units   
4                          Imperviousness   
5                              Land_Cover   
6                                Land_Use   
7                               Landcover   
8                      Population_Density   
9                        Precipitation_mm   
10                                 Runoff   
11                       Slope_Categories   
12             Total_Copper_Concentration   
13  Total_Kjeldahl_Nitrogen_Concentration   
14         Total_Phosphorus_Concentration   
15   Total_Suspended_Solids_Concentration   
16                                Traffic   

                              pretty_name  \
0                   Age of Imperviousness   
1                     Flow Duration Index   
2                    HSPF Land Cover Type   
3               Hydrologic Response Units   
4                          Imperviousness   
5                              Land Cover   
6                                Land Use   
7                               Landcover   
8                      Population Density   
9                        Precipitation mm   
10                                 Runoff   
11                       Slope Categories   
12             Total Copper Concentration   
13  Total Kjeldahl Nitrogen Concentration   
14         Total Phosphorus Concentration   
15   Total Suspended Solids Concentration   
16                                Traffic   

                                            json_data  
0   {"discrete": "TRUE", "sourceName": "The Nature...  
1   {"discrete": "FALSE", "sourceName": "The Natur...  
2   {"discrete": "TRUE", "sourceName": "The Nature...  
3   {"discrete": "TRUE", "sourceName": "The Nature...  
4   {"discrete": "FALSE", "sourceName": "The Natur...  
5   {"discrete": "TRUE", "source

In [44]:
import requests

def send_request(session, asset_name,props):
    # Earth Engine enabled Cloud Project.
    project_folder = 'ee-swhm'
    # A folder (or ImageCollection) name and the new asset name.
    asset_id = asset_name

    #url = 'https://earthengine.googleapis.com/v1alpha/projects/{}/assets?assetId=production_layers/{}'
    url = 'https://earthengine.googleapis.com/v1alpha/projects/{}/assets?assetId=production_layers/{}'
    params = {'overwrite': True}
    response = session.post(
      url = url.format(project_folder, asset_id),
      data = props, 
        params=params
    )
    
    print(json.loads(response.content))
    
    


In [45]:
import json
import pprint
def process_df(df): 
    for index, row in df.iterrows():
        json_string = row['json_data']
        print(row['pretty_name'])
        print('...')
        if json_string is not None and json_string != 'NA':
            layer = df['pretty_name'][index]
            asset_name = df['asset_name'][index]
            uri = df['uri'][index]

            # Request body as a dictionary.
            request = {
                'type': 'IMAGE',
                'gcs_location': {'uris': uri},
                'properties':json.loads(df['json_data'][index])
            }

            props = json.dumps(request)
            send_request(session, asset_name,props)


In [48]:
df.iloc[10:11]

blob_name                                       uri asset_name  \
10  rasters/Runoff.tif  gs://live_data_layers/rasters/Runoff.tif     Runoff   

   pretty_name                                          json_data  
10      Runoff  {"discrete": "FALSE", "sourceName": "The Natur...

## Send the request

Make the POST request to the Earth Engine [`projects.assets.create`](https://developers.google.com/earth-engine/reference/rest/v1alpha/projects.assets/create) endpoint.

In [49]:
process_df(df.iloc[10:11])

Runoff
...
{'type': 'IMAGE', 'name': 'projects/ee-swhm/assets/production_layers/Runoff', 'id': 'projects/ee-swhm/assets/production_layers/Runoff', 'properties': {'default_reduction': 'mean', 'description': 'Mean annual runoff calculated through continuous simulation for the period 1970-1999.', 'discrete': 'FALSE', 'pretty_name': 'Runoff', 'safe_name': 'runoff_mm', 'scale': 10, 'sourceName': 'The Nature Conservancy', 'units': 'mm/year'}, 'updateTime': '2023-03-30T20:46:56.489983Z', 'geometry': {'type': 'LineString', 'coordinates': [[-122.01511824692012, 46.59292817722988], [-121.68327206144858, 46.59283220447431], [-121.3514260605525, 46.59292814410767], [-121.01958003287811, 46.59283222497405], [-120.65446980545009, 46.5928848791753], [-120.65446551178702, 49.31345857490763], [-121.45097994583746, 49.31345863174228], [-122.38014887572923, 49.31345858360832], [-123.11021021151346, 49.31345862985674], [-123.64116391851715, 49.31345864332467], [-124.43759441590882, 49.313458611434946], [-

PATCH https://earthengine.googleapis.com/v1/{asset.name=projects/*/assets/**}

## Details on COG-backed assets

### Permissions
The ACLs of COG-backed Earth Engine assets and the underlying data are managed separately. If a COG-backed asset is shared in Earth Engine, it is the owner's responsibility to ensure that the data in GCS is shared with the same parties. If the data is not visible, Earth Engine will return an error of the form "Failed to load the GeoTIFF at `gs://my-bucket/my-object#123456`" (123456 is the generation of the object).

### Generations
When a COG-backed asset is created, Earth Engine reads the metadata of the TIFF in Cloud Storage and creates asset store entry. The URI associated with that entry must have a generation.  See the [object versioning docs](https://cloud.google.com/storage/docs/object-versioning) for details on generations. If a generation is specified (e.g., `gs://foo/bar#123`), Earth Engine will use it. If a generation is not specified, Earth Engine will use the latest generation of the object. 

That means that if the object in GCS is updated, Earth Engine will return a "Failed to load the GeoTIFF at `gs://my-bucket/my-object#123456`" error because the expected object no longer exists (unless the bucket enables  multiple object versions).  This policy is designed to keep metadata of the asset in sync with the metadata of the object.  

### Configuration
In terms of how a COG should be configured, the TIFF MUST be:

- Tiled, where the tile dimensions are either:
  - 16x16
  - 32x32
  - 64x64
  - 128x128
  - 256x256
  - 512x512
  - 1024x1024

- Arranged so that all IFDs are at the beginning.

For best performance:

- Use tile dimensions of 128x128 or 256x256.
- Include power of 2 overviews.

See [this page](https://cogeotiff.github.io/rio-cogeo/Advanced/#web-optimized-cog) for more details on an optimized configuration.